src/server/pkg/work/work.go

```diff
@@ -22,21 +22,18 @@ type CollectFunc func(context.Context, *Task) error
 
 // Master is the master for a task.
 // The master will layout the subtasks for the task in etcd and collect
-// them upon completion. The collectFunc callback will be called for
-// each subtask that is collected.
+// them upon completion.
 type Master struct {
 	etcdClient          *etcd.Client
 	taskCol, subtaskCol col.Collection
-	collectFunc         CollectFunc
 }
 
 // NewMaster creates a new master.
-func NewMaster(etcdClient *etcd.Client, etcdPrefix string, collectFunc CollectFunc) *Master {
+func NewMaster(etcdClient *etcd.Client, etcdPrefix string, taskNamespace string) *Master {
```

Comment:
A couple changes to the interface here:
1. tasks are now namespaced by an extra path component in etcd, this is so that we can have different task types using this interface, otherwise the worker layer might receive a storage layer task
2. the callbacks are now provided in the `Run` function rather than the constructor, this lets you create the master or worker objects before you have everything in scope to run a task.

Reply:
Definitely agree with the first change. I am lukewarm about the second. It seems like we are disconnecting a task from the functionality of the master/worker in the task. This is more flexible, but seems like it could also open the door to problems. I do see why you would prefer this if you want/need state that is in scope of the run function, but not the constructor.